src/hotspot/share/gc/g1/g1YoungCollector.cpp

```diff
@@ -535,6 +536,12 @@ void G1YoungCollector::pre_evacuate_collection_set(G1EvacInfo* evacuation_info)
 #endif
 
   allocation_failure_injector()->arm_if_needed();
+
+#ifdef ASSERT
+  for (JavaThreadIteratorWithHandle jtiwh; JavaThread *thread = jtiwh.next(); ) {
+    assert(G1ThreadLocalData::pin_count_cache(thread).count() == 0, "must be flushed");
+  }
+#endif
```

Comment:
It's not obvious to see this assert in this context. I wonder if it's cleaner (or more natural) if it's moved to `JavaThreadRetireTLABAndFlushLogs`, where we flush the cache.

Reply:
It was meant similarly to the call to `_g1h->verifier()->check_region_attribute_table()` call to verify post conditions for that GC phase.
Putting it right after `JavaThreadRetireTLABAndFlushLogs` seems too close to the change to kind of make it kind of redundant - i.e. that task is small enough to understand it at a glance anyway (well, me at least).

Thinking again, it might be most useful to put this into a method (like mentioned `check_region_attribute_table`) and call it there to be able to use it elsewhere during debugging if needed.

Of course, we can also just remove this - what do you think?

Reply:
Removing it is better, IMO.

Reply:
Removed the super-cautious assert.